### PR TITLE
Added support for schema inference for `hdfsCluster`

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
Changelog category (leave one):

- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added support for schema inference for `hdfsCluster`.

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
Also added [CLI MiniCluster](https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/CLIMiniCluster.html) to the testing container to improve testing of that functions

File: src/Storages/HDFS/StorageHDFSCluster.cpp [hunk 1]
```diff
@@ -41,10 +48,23 @@ StorageHDFSCluster::StorageHDFSCluster(
...
+    if (columns_.empty())
+    {
+        auto columns = StorageHDFS::getTableStructureFromData(format_name, uri_, compression_method, context_);
+        storage_metadata.setColumns(columns);
+    }
+    else
+        storage_metadata.setColumns(columns_);
+
     storage_metadata.setConstraints(constraints_);
-    setInMemoryMetadata(storage_metadata);
```
Comment (Avogar) [inline src/Storages/HDFS/StorageHDFSCluster.cpp:47]: Why did you delete it?

File: tests/queries/0_stateless/02244_hdfs_cluster.sql [hunk 2]
```diff
@@ -0,0 +1,12 @@
+-- Tags: no-fasttest
+-- Tag no-fasttest: Depends on Java
+
+insert into table function hdfs('hdfs://localhost:12222/test_1', 'TSV', 'column1 UInt32, column2 UInt32, column3 UInt32') select 1, 2, 3 settings hdfs_truncate_on_insert=1;
+insert into table function hdfs('hdfs://localhost:12222/test_2', 'TSV', 'column1 UInt32, column2 UInt32, column3 UInt32') select 4, 5, 6 settings hdfs_truncate_on_insert=1;
+insert into table function hdfs('hdfs://localhost:12222/test_3', 'TSV', 'column1 UInt32, column2 UInt32, column3 UInt32') select 7, 8, 9 settings hdfs_truncate_on_insert=1;
+
+select * from hdfs('test_cluster_two_shards_localhost', 'hdfs://localhost:12222/test_{1,2,3}');
```
Comment (Avogar) [inline tests/queries/0_stateless/02244_hdfs_cluster.sql:8]: Let's also add tests with describe query

File: tests/queries/0_stateless/02244_hdfs_cluster.sql [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+select * from hdfs('test_cluster_two_shards_localhost', 'hdfs://localhost:12222/test_{1,2,3}', 'TSV', 'column1 UInt32, column2 UInt32, column3 UInt32');
+
+select * from hdfsCluster('test_cluster_two_shards_localhost', 'hdfs://localhost:12222/test_{1,2,3}');
```
Comment (Avogar) [inline tests/queries/0_stateless/02244_hdfs_cluster.sql:11]: Let's also add tests with describe query

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

**Review by Avogar — APPROVED** [review]

**Comment by nikitamikhaylov** [comment]
Flaky check - timeout..
